sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/SchemaAndRowProto.java

```diff
@@ -0,0 +1,131 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.io.gcp.bigquery;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.api.services.bigquery.model.TableRow;
+import com.google.cloud.bigquery.v3.RowOuterClass.Row;
+import com.google.cloud.bigquery.v3.RowOuterClass.StructField;
+import com.google.cloud.bigquery.v3.RowOuterClass.StructType;
+import com.google.cloud.bigquery.v3.RowOuterClass.StructValue;
+import com.google.cloud.bigquery.v3.RowOuterClass.TypeKind;
+import com.google.cloud.bigquery.v3.RowOuterClass.Value;
+import com.google.cloud.bigquery.v3.RowOuterClass.Value.ValueCase;
+import javax.annotation.Nullable;
+
+/**
+ * A wrapper for a {@link Row} and the {@link StructType} representing the schema of the
+ * table it was generated from. Also provides utility methods to extract data.
+ */
+public class SchemaAndRowProto {
+  private final StructType tableSchema;
+  private final Row row;
+  TableRow tableRow;
+
+  public SchemaAndRowProto(StructType tableSchema, Row row) {
+    this.tableSchema = checkNotNull(tableSchema, "tableSchema");
+    this.row = checkNotNull(row, "row");
+  }
+
+  public Row getRow() {
+    return row;
+  }
+
+  public StructType getSchema() {
+    return tableSchema;
+  }
+
+  /**
+   * Gets the value of the specified (possibly nested) field or null if it can't be found.
+   */
+  @Nullable
+  public Object get(String fieldName) {
+    String[] subFieldNames = fieldName.split("\\.");
+    StructType structType = tableSchema;
+    StructValue structValue = row.getValue();
+
+    for (int i = 0; i < subFieldNames.length - 1; i++) {
+      int fieldIndex = getFieldIndex(structType, subFieldNames[i]);
+      if (fieldIndex == -1) {
+        return null;
+      }
+      StructField structField = structType.getFields(fieldIndex);
+      if (structField.getFieldType().getTypeKind() != TypeKind.TYPE_STRUCT) {
+        return null;
+      }
+
+      Value fieldValue = structValue.getFields(fieldIndex);
+      if (fieldValue.getValueCase() == ValueCase.NULL_VALUE) {
+        return null;
+      }
+
+      structType = structField.getFieldType().getStructType();
+      structValue = fieldValue.getStructValue();
+    }
+
+    String leafFieldName = subFieldNames[subFieldNames.length - 1];
+    int fieldIndex = getFieldIndex(structType, leafFieldName);
+    if (fieldIndex == -1) {
+      return null;
+    }
+
+    StructField leafField = structType.getFields(fieldIndex);
+    Value fieldValue = structValue.getFields(fieldIndex);
+    if (fieldValue.getValueCase() == ValueCase.NULL_VALUE) {
+      return null;
+    }
+
+    switch (leafField.getFieldType().getTypeKind()) {
```

Comment:
Seems like this method should be in BigQueryRowProtoUtils, it seems tightly related to other stuff there, and it needs 100% test coverage also.

Reply:
BigQueryRowProtoUtils actually contains only methods related to Row proto -> TableRow conversion, as is the case with BigQueryAvroUtils today. I can move this code into the utils class, however.